### ISSUE TITLE
Misc. bug: llama-server, speed penalty from d9d398f since b5222

### ISSUE THREAD

**Issue by cmhamiche** [ggml-org/llama.cpp#15672]
### Name and Version

version: 6319 (792b44f)
built with MSVC 19.44.35215.0 for x64
cmake -B build -DCMAKE_BUILD_TYPE=Release -DGGML_CUDA=ON -DCMAKE_CUDA_ARCHITECTURES="86" -DLLAMA_CURL=OFF

Intel i3-12100F, 2*32GB 3200 Mhz, Nvidia RTX 3060 12 GB

### Operating systems

Windows

### Problem description & steps to reproduce

With **llama-server**, there's a noticeable penalty from d9d398f since b5222 :
"sampling : when top-k <= 0 -> noop (#13173)"
It's a single line in src/llama-sampling.cpp, **llama-cli does not seem to be affected**

### First Bad Commit

There's comparable differences between llama-b5221-bin-win-cuda-cu12.4-x64 and llama-b5222-bin-win-cuda-cu12.4-x64

### Relevant log output
-------------------------------------------
#### llama-server _without cpu offloading_: top-k 0.1, min-p 0.1
gemma-2-2b-it.F16.gguf

llama-server.exe --no-mmap -t 7 -fa -ngl 99 -c 8000 -n 1000 --temp 1 --top-p 1 --top-k 0.1 --min-p 0.1 --jinja -m gemma-2-2b-it.F16.gguf
[...]
load_tensors: offloaded 27/27 layers to GPU
load_tensors:    CUDA_Host model buffer size =  1125.00 MiB
load_tensors:        CUDA0 model buffer size =  4986.92 MiB

**b6319 with reverted d9d398f, "sampling : when top-k <= 0 -> noop (#13173)"**
prompt eval time =     470.11 ms /   453 tokens (    1.04 ms per token,   963.61 tokens per second)
       eval time =   16651.99 ms /   476 tokens (   34.98 ms per token,    28.59 tokens per second)
      total time =   17122.10 ms /   929 tokens

**b6319 default**
prompt eval time =     388.79 ms /   453 tokens (    0.86 ms per token,  1165.15 tokens per second)
       eval time =   18129.15 ms /   458 tokens (   39.58 ms per token,    25.26 tokens per second)
      total time =   18517.94 ms /   911 tokens

-------------------------------------------
#### llama-server _with cpu offloading_: ncmoe 32, top-k 0, min-p 0
gpt-oss-120b-F16.gguf

llama-server.exe --no-mmap -t 7 -fa -ngl 99 -ncmoe 32 -c 30000 -n 1000 --temp 1 --top-p 1 --top-k 0 --min-p 0 --jinja -m gpt-oss-120b-F16.gguf
[...]
load_tensors: offloaded 37/37 layers to GPU
load_tensors:          CPU model buffer size = 52877.11 MiB
load_tensors:        CUDA0 model buffer size =  9451.23 MiB

**b6319 with reverted d9d398f, "sampling : when top-k <= 0 -> noop (#13173)"**
prompt eval time =    5145.30 ms /   495 tokens (   10.39 ms per token,    96.20 tokens per second)
       eval time =  101352.60 ms /  1000 tokens (  101.35 ms per token,     9.87 tokens per second)
      total time =  106497.90 ms /  1495 tokens

**b6319 default**
prompt eval time =    4994.36 ms /   495 tokens (   10.09 ms per token,    99.11 tokens per second)
       eval time =  105848.31 ms /  1000 tokens (  105.85 ms per token,     9.45 tokens per second)
      total time =  110842.66 ms /  1495 tokens

-------------------------------------------
#### llama-server _with some shared ram_: top-k 0.1, min-p 0.1
Qwen3-14B-Q4_K_M.gguf 

llama-server.exe --no-mmap -t 7 -fa -ngl 99 -c 8000 -n 1000 --temp 1 --top-p 1 --top-k 0.1 --min-p 0.1 --jinja -m Qwen3-14B-Q4_K_M.gguf
[...]
load_tensors: offloaded 41/41 layers to GPU
load_tensors:        CUDA0 model buffer size =  8161.75 MiB
load_tensors:          CPU model buffer size =   417.30 MiB

**b6319 with reverted d9d398f, "sampling : when top-k <= 0 -> noop (#13173)"**
prompt eval time =     864.52 ms /   518 tokens (    1.67 ms per token,   599.18 tokens per second)
       eval time =   38920.46 ms /  1000 tokens (   38.92 ms per token,    25.69 tokens per second)
      total time =   39784.98 ms /  1518 tokens

**b6319 default**
prompt eval time =     803.84 ms /   518 tokens (    1.55 ms per token,   644.40 tokens per second)
       eval time =   42078.33 ms /  1000 tokens (   42.08 ms per token,    23.77 tokens per second)
      total time =   42882.18 ms /  1518 tokens

-------------------------------------------
#### CLI for comparison, llama-cli _with some shared ram_: top-k 0, min-p 0
**Qwen3-14B-Q4_K_M.gguf** 

llama-cli.exe --no-mmap -t 7 -fa -ngl 99 -c 8000 -n 1000 --temp 1 --top-p 1 --top-k 0 --min-p 0 --jinja -m Qwen3-14B-Q4_K_M.gguf -p "On a typical 24-hour digital alarm clock, what time will the sum of the four digits be the highest?"
[...]
load_tensors: offloaded 41/41 layers to GPU
load_tensors:        CUDA0 model buffer size =  8161.75 MiB
load_tensors:          CPU model buffer size =   417.30 MiB

**b6319 with reverted d9d398f, "sampling : when top-k <= 0 -> noop (#13173)"**
llama_perf_sampler_print:    sampling time =   10327.16 ms /  1032 runs   (   10.01 ms per token,    99.93 tokens per second)
llama_perf_context_print:        load time =    2430.43 ms
llama_perf_context_print: prompt eval time =      61.37 ms /    32 tokens (    1.92 ms per token,   521.44 tokens per second)
llama_perf_context_print:        eval time =   27465.81 ms /   999 runs   (   27.49 ms per token,    36.37 tokens per second)
llama_perf_context_print:       total time =   40954.06 ms /  1031 tokens
llama_perf_context_print:    graphs reused =        995

**b6319 default**
llama_perf_sampler_print:    sampling time =   13363.90 ms /  1032 runs   (   12.95 ms per token,    77.22 tokens per second)
llama_perf_context_print:        load time =    2433.27 ms
llama_perf_context_print: prompt eval time =      60.15 ms /    32 tokens (    1.88 ms per token,   532.05 tokens per second)
llama_perf_context_print:        eval time =   27404.76 ms /   999 runs   (   27.43 ms per token,    36.45 tokens per second)
llama_perf_context_print:       total time =   41827.05 ms /  1031 tokens
llama_perf_context_print:    graphs reused =        995

**Comment by alankila** [comment]
I took oprofile trace of debug build of llama-cli. I noticed that llama_sampler_softmax_impl was at 9 % cpu according to the trace.

I noticed that multiple samplers seem to perform the softmax transformation. It is done a whole bunch of times, so maybe it could be investigated if it could be put behind a gate so that it isn't applied repeatedly if nothing is different, or something, maybe? The softmax function doesn't use much CPU if you set any --top-k, probably because it's faster to do when the vocabulary for the sampler gets much smaller.

Edit: I made a mistake in my first version of comment, as I originally thought it overwrote the .logit itself with softmax version which would have been a correctness issue. That's not actually a problem. It is fine to call it multiple times, as the function writes to .p from .logit, except that it seems to be slow.

**Comment by cmhamiche** [comment]
OK, nevermind. I was either mistaken or the latest changes are incredible:

load_tensors: offloaded 27/27 layers to GPU
load_tensors:    CUDA_Host model buffer size =  1125.00 MiB
load_tensors:        CUDA0 model buffer size =  4986.92 MiB

-----------------------

**version: 6338 (e92d53b2) without d9d398f: "sampling : when top-k <= 0 -> noop (#13173)"**

PS C:\Users\Windows\Programmes\llamacpp\llama.cpp_no_fix\build\bin\RelWithDebInfo> nsys profile --trace=cuda,nvtx ./llama-server.exe --no-mmap -t 7 -ngl 99 -fa on -c 8000 -n 1000 --temp 1 **--top-k 0 --min-p 0.1** --jinja -m "C:\Users\Windows\Programmes\llamacpp\models\gemma-2-2b-it.F16.gguf"

prompt eval time =     456.86 ms /   485 tokens (    0.94 ms per token,  1061.60 tokens per second)
       eval time =   27815.05 ms /   592 tokens (   46.98 ms per token,    **21.28** tokens per second)
      total time =   28271.91 ms /  1077 tokens


**version: 6338 (e92d53b2)**

PS C:\Users\Windows\Programmes\llamacpp\llama.cpp\build\bin\RelWithDebInfo> nsys profile --trace=cuda,nvtx ./llama-server.exe --no-mmap -t 7 -ngl 99 -fa on -c 8000 -n 1000 --temp 1 **--top-k 0 --min-p 0.1** --jinja -m "C:\Users\Windows\Programmes\llamacpp\models\gemma-2-2b-it.F16.gguf"

prompt eval time =     487.56 ms /   485 tokens (    1.01 ms per token,   994.75 tokens per second)
       eval time =    9786.45 ms /   449 tokens (   21.80 ms per token,    **45.88** tokens per second)
      total time =   10274.01 ms /   934 tokens

---------------------

**version: 6338 (e92d53b2) without d9d398f: "sampling : when top-k <= 0 -> noop (#13173)"**

PS C:\Users\Windows\Programmes\llamacpp\llama.cpp_no_fix\build\bin\RelWithDebInfo> nsys profile --trace=cuda,nvtx ./llama-server.exe --no-mmap -t 7 -ngl 99 -fa on -c 8000 -n 1000 --temp 1 **--top-k 0.1 --min-p 0.1** --jinja -m "C:\Users\Windows\Programmes\llamacpp\models\gemma-2-2b-it.F16.gguf"

prompt eval time =     388.84 ms /   485 tokens (    0.80 ms per token,  1247.31 tokens per second)
       eval time =   26481.33 ms /   759 tokens (   34.89 ms per token,    **28.66** tokens per second)
      total time =   26870.17 ms /  1244 tokens


**version: 6338 (e92d53b2)**

PS C:\Users\Windows\Programmes\llamacpp\llama.cpp\build\bin\RelWithDebInfo> nsys profile --trace=cuda,nvtx ./llama-server.exe --no-mmap -t 7 -ngl 99 -fa on -c 8000 -n 1000 --temp 1 **--top-k 0.1 --min-p 0.1** --jinja -m "C:\Users\Windows\Programmes\llamacpp\models\gemma-2-2b-it.F16.gguf"

prompt eval time =     396.37 ms /   485 tokens (    0.82 ms per token,  1223.60 tokens per second)
       eval time =   10011.61 ms /   459 tokens (   21.81 ms per token,    **45.85** tokens per second)
      total time =   10407.98 ms /   944 tokens

-----------------------

**version: 6338 (e92d53b2) without d9d398f: "sampling : when top-k <= 0 -> noop (#13173)"**

PS C:\Users\Windows\Programmes\llamacpp\llama.cpp_no_fix\build\bin\RelWithDebInfo> nsys profile --trace=cuda,nvtx ./llama-server.exe --no-mmap -t 7 -ngl 99 -fa on -c 8000 -n 1000 --temp 1 **--top-k 0 --min-p 0** --jinja -m "C:\Users\Windows\Programmes\llamacpp\models\gemma-2-2b-it.F16.gguf"

prompt eval time =     330.31 ms /   485 tokens (    0.68 ms per token,  1468.34 tokens per second)
       eval time =   24906.48 ms /   717 tokens (   34.74 ms per token,    **28.79** tokens per second)
      total time =   25236.79 ms /  1202 tokens


**version: 6338 (e92d53b2)**

PS C:\Users\Windows\Programmes\llamacpp\llama.cpp\build\bin\RelWithDebInfo> nsys profile --trace=cuda,nvtx ./llama-server.exe --no-mmap -t 7 -ngl 99 -fa on -c 8000 -n 1000 --temp 1 **--top-k 0 --min-p 0** --jinja -m "C:\Users\Windows\Programmes\llamacpp\models\gemma-2-2b-it.F16.gguf"

prompt eval time =     384.13 ms /   485 tokens (    0.79 ms per token,  1262.61 tokens per second)
       eval time =   20986.98 ms /   958 tokens (   21.91 ms per token,    **45.65** tokens per second)
      total time =   21371.11 ms /  1443 tokens

------------------

**Sanity check, \llama-b6335-bin-win-cuda-12.4-x64**

./llama-server.exe --no-mmap -t 7 -ngl 99 -fa on -c 8000 -n 1000 --temp 1 --top-k 0 --min-p 0.1 --jinja -m "C:\Users\Windows\Programmes\llamacpp\models\gemma-2-2b-it.F16.gguf"

prompt eval time =     390.51 ms /     1 tokens (  390.51 ms per token,     2.56 tokens per second)
       eval time =   30917.91 ms /   787 tokens (   39.29 ms per token,    **25.45** tokens per second)
      total time =   31308.42 ms /   788 tokens

------------------

I think there's a lot going with the samplers right now, feel free to close this.

edit: even with llama-server (-fa on -ncmoe 32) and gpt-oss-120b-F16.gguf on my modest machine, I see a big increase in gpu utilization, something like 15 to 20% and ofc that correlates with the tokens per seconds.
